doc/man/heketi-cli.8
.\"
.\" Copyright (c) 2015 The heketi Authors
.\"
.\" This file is licensed to you under your choice of the GNU Lesser
.\" General Public License, version 3 or any later version (LGPLv3 or
.\" later), or the GNU General Public License, version 2 (GPLv2), in all
.\" cases as published by the Free Software Foundation.
.\"

.TH heketi-cli 8 "Heketi command line program" "Apr 2016" "The heketi Authors"
.nh
.ad l
.SH NAME
.PP
heketi\-cli \- Command line program for Heketi
.SH SYNOPSIS
.PP
\fBheketi\-cli\fP [commands] [options]
.SH DESCRIPTION
.PP
Command line program for Heketi
.SH COMMANDS


.SS "Cluster Commands"
.PP
.TP

\fBheketi\-cli cluster create\fP
Create a cluster
    \fBExample\fP
    $ heketi-cli cluster create
.PP
.TP

\fBheketi\-cli cluster delete <CLUSTER-ID>\fP
Delete a cluster
    \fBExample\fP
    $ heketi-cli cluster delete 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli cluster info  <CLUSTER-ID>\fP
Retrieves information about cluster

    \fBExample\fP
    $ heketi-cli cluster info 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli cluster list\fP
Lists the clusters managed by Heketi

    \fBExample\fP
    $ heketi-cli cluster list


.SS "Device Commands"
.PP
.TP

\fBheketi\-cli device add \-\-name=<DEVICE-NAME> \-\-node=<NODE-ID>\fP
Add new device to node to be managed by Heketi
.TP
\fB           Options\fP
.PP
\fB               \-\-name\fP=""
                   Name of device to add
.PP
\fB               \-\-node\fP=""
                   Id of the node which has this device

    \fB       Example\fP
           $ heketi\-cli device add \\
               \-\-name=/dev/sdb
               \-\-node=3e098cb4407d7109806bb196d9e8f095

.PP
.TP

\fBheketi\-cli device delete <DEVICE-ID>\fP
Deletes a device from Heketi node

    \fBExample\fP
    $ heketi-cli device delete 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli device disable <DEVICE\-ID>\fP
Disallow usage of a device by placing it offline

    \fBExample\fP
    $ heketi\-cli device disable 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli device enable <DEVICE\-ID>\fP
Allows device to go online

    \fBExample\fP
    $ heketi\-cli device enable 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli device info  <DEVICE-ID>\fP
Retrieves information about device

    \fBExample\fP
    $ heketi-cli device info 886a86a868711bef83001


.SS "Node Commands"
.PP
.TP

\fBheketi\-cli node add \-\-zone=<ZONE-NUMBER> \-\-cluster=<CLUSTER-ID> \-\-management\-host\-name=<MANAGEMENT-HOSTNAME> \-\-storage-host-name=<STORAGE-HOSTNAME>\fP
Add new node to be managed by Heketi
.TP
\fB           Options\fP
.PP
\fB               \-\-cluster\fP=""
                   The cluster in which the node should reside
.PP
\fB               \-\-management\-host\-name\fP=""
                   Management host name
.PP
\fB               \-\-storage\-host\-name\fP=""
                   Storage host name
.PP
\fB               \-\-zone\fP=\-1
                   The zone in which the node should reside

    \fB       Example\fP
           $ heketi\-cli node add \\
               \-\-zone=3 \\
               \-\-cluster=3e098cb4407d7109806bb196d9e8f095 \\
               \-\-management\-host\-name=node1\-manage.gluster.lab.com \\
               \-\-storage\-host\-name=node1\-storage.gluster.lab.com

.PP
.TP

\fBheketi\-cli node delete <NODE-ID>\fP
Deletes a node from Heketi management

    \fBExample\fP
    $ heketi-cli node delete 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli node disable <NODE\-ID>\fP
Disallow usage of a node by placing it offline

    \fBExample\fP
    $ heketi\-cli node disable 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli node enable <NODE\-ID>\fP
Allows node to go online

    \fBExample\fP
    $ heketi\-cli node enable 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli node info  <NODE-ID>\fP
Retrieves information about node

    \fBExample\fP
    $ heketi-cli node info 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli node list\fP
List all nodes in cluster

    \fBExample\fP
    $ heketi\-cli node list



.SS "Setup OpenShift/Kubernetes persistent storage for Heketi"
.PP
.TP

\fBheketi\-cli setup\-openshift\-heketi\-storage\fP
Creates a dedicated GlusterFS volume for Heketi.
Once the volume is created, a Kubernetes/OpenShift
list object is created to configure the volume.
.TP
\fB           Options\fP
.PP
\fB               \-\-listfile\fP="heketi\-storage.json"
                   Filename to contain list of objects

\fB               \-\-durability\fP="replicate"
                   Durability of the Heketi database storage

    \fB       Example\fP
           $ heketi\-cli setup\-openshift\-heketi\-storage

.PP



.SS "Topology Commands"
.PP
.TP

\fBheketi\-cli topology load \-\-json=<JSON-FILENAME>\fP
Add devices to Heketi from a configuration file
.TP
\fB           Options\fP
.PP
\fB               \-j, \-\-json\fP=""
                   Configuration containing devices, nodes, and clusters, in JSON format

    \fB       Example\fP
           $ heketi-cli topology load --json=topo.json

.PP
.TP

\fBheketi\-cli topology info \fP
Retreives information about the current Topology

    \fBExample\fP
    $ heketi-cli topology info


.SS "Volume Commands"
.PP
.TP

\fBheketi\-cli volume create \-\-clusters=<CLUSTER-IDS> \-\-disperse-data=<DISPERSION-VALUE> \-\-durability=<TYPE> \-\-name=<VOLUME-NAME> \-\-redundancy=<REDUNDENCY-VALUE> \-\-replica=<REPLICA-VALUE> \-\-size=<VOLUME-SIZE> \-\-snapshot-factor=<SNAPSHOT-FACTOR-VALUE>\fP
Create a GlusterFS volume
.TP
\fB           Options\fP
.PP
\fB               \-\-clusters\fP=""
.PP
.RS
.nf
           Optional: Comma separated list of cluster ids where this volume
           must be allocated. If omitted, Heketi will allocate the volume
           on any of the configured clusters which have the available space.
           Providing a set of clusters will ensure Heketi allocates storage
           for this volume only in the clusters specified.
.fi
.RE
.PP
\fB               \-\-disperse\-data\fP=4
.PP
.RS
.nf
           Optional: Dispersion value for durability type 'disperse'.
           Default is 4
.fi
.RE
.PP
\fB               \-\-durability\fP="replicate"
.PP
.RS
.nf
           Optional: Durability type.
                     Values are:
                     none: No durability. Distributed volume only.
                     replicate: (Default) Distributed\-Replica volume.
                     disperse: Distributed\-Erasure Coded volume.
.fi
.RE
.PP
\fB               \-\-gid\fP=0
.PP
.RS
.nf
            Optional: Initialize volume with the specified group id
            Default is 0
.fi
.RE
.PP
\fB               \-\-name\fP=""
.PP
.RS
.nf
            Optional: Name of volume. Only set if really necessary
.fi
.RE
.PP
\fB               \-\-persistent\-volume\fP[=false]
.PP
.RS
.nf
            Optional: Output to standard out a persistent volume JSON file for OpenShift or
            Kubernetes with the name provided.
.fi
.RE
.PP
\fB               \-\-persistent\-volume\-endpoint\fP=""
.PP
.RS
.nf
            Optional: Endpoint name for the persistent volume
.fi
.RE
.PP
\fB               \-\-persistent\-volume\-file\fP=""
.PP
.RS
.nf
            Optional: Create a persistent volume JSON file for OpenShift or
            Kubernetes with the name provided.
.fi
.RE
.PP
\fB               \-\-redundancy\fP=2
.PP
.RS
.nf
            Optional: Redundancy value for durability type 'disperse'.
            Default is 2
.fi
.RE
.PP
\fB               \-\-replica\fP=3
.PP
.RS
.nf
            Replica value for durability type 'replicate'.
            Default is 3
.fi
.RE
.PP
\fB               \-\-size\fP=\-1
.PP
.RS
.nf
            Size of volume in GiB
.fi
.RE
.PP
\fB               \-\-snapshot\-factor\fP=1
.PP
.RS
.nf
            Optional: Amount of storage to allocate for snapshot support.
            Must be greater 1.0.  For example if a 10TiB volume requires 5TiB of
            snapshot storage, then snapshot\-factor would be set to 1.5.  If the
            value is set to 1, then snapshots will not be enabled for this volume
.fi
.RE


    \fB       Example\fP
           * Create a 100GiB replica 3 volume:
                 $ heketi\-cli volume create \-\-size=100
           * Create a 100GiB replica 3 volume specifying two specific clusters:
                 $ heketi\-cli volume create \-\-size=100 \\
                 \-\-clusters=0995098e1284ddccb46c7752d142c832,60d46d518074b13a04ce1022c8c7193c
           * Create a 100GiB replica 2 volume with 50GiB of snapshot storage:
                 $ heketi\-cli volume create \-\-size=100 \-\-snapshot\-factor=1.5 \-\-replica=2
           * Create a 100GiB distributed volume
                 $ heketi\-cli volume create \-\-size=100 \-\-durability=none
           * Create a 100GiB erasure coded 4+2 volume with 25GiB snapshot storage:
                 $ heketi\-cli volume create \-\-size=100 \-\-durability=disperse \-\-snapshot\-factor=1.25
           * Create a 100GiB erasure coded 8+3 volume with 25GiB snapshot storage:
                 $ heketi\-cli volume create \-\-size=100 \-\-durability=disperse \-\-snapshot\-factor=1.25 \\
                 \-\-disperse\-data=8 \-\-redundancy=3


.PP
.TP

\fBheketi\-cli volume delete <VOLUME-ID>\fP
Deletes the volume

    \fBExample\fP
    $ heketi-cli volume delete 886a86a868711bef83001
.PP
.TP

\fBheketi\-cli volume expand --expand-size=<SIZE> --volume=<VOLUME-ID>\fP
Expand a volume
.TP
\fB           Options\fP
.PP
\fB               \-\-expand\fP=""
                   Amount in GiB to add to the volume
.PP
\fB               \-\-volume\fP=""
                    Id of volume to expand


\fB           Example\fP
               * Add 10GiB to a volume

client/cli/go/heketi-cli.sh
#!/bin/bash

# bash completion for heketi-cli                           -*- shell-script -*-

__debug()
{
    if [[ -n ${BASH_COMP_DEBUG_FILE} ]]; then
        echo "$*" >> "${BASH_COMP_DEBUG_FILE}"
    fi
}

# Homebrew on Macs have version 1.3 of bash-completion which doesn't include
# _init_completion. This is a very minimal version of that function.
__my_init_completion()
{
    COMPREPLY=()
    _get_comp_words_by_ref "$@" cur prev words cword
}

__index_of_word()
{
    local w word=$1
    shift
    index=0
    for w in "$@"; do
        [[ $w = "$word" ]] && return
        index=$((index+1))
    done
    index=-1
}

__contains_word()
{
    local w word=$1; shift
    for w in "$@"; do
        [[ $w = "$word" ]] && return
    done
    return 1
}

__handle_reply()
{
    __debug "${FUNCNAME[0]}"
    case $cur in
        -*)
            if [[ $(type -t compopt) = "builtin" ]]; then
                compopt -o nospace
            fi
            local allflags
            if [ ${#must_have_one_flag[@]} -ne 0 ]; then
                allflags=("${must_have_one_flag[@]}")
            else
                allflags=("${flags[*]} ${two_word_flags[*]}")
            fi
            COMPREPLY=( $(compgen -W "${allflags[*]}" -- "$cur") )
            if [[ $(type -t compopt) = "builtin" ]]; then
                [[ "${COMPREPLY[0]}" == *= ]] || compopt +o nospace
            fi

            # complete after --flag=abc
            if [[ $cur == *=* ]]; then
                if [[ $(type -t compopt) = "builtin" ]]; then
                    compopt +o nospace
                fi

                local index flag
                flag="${cur%%=*}"
                __index_of_word "${flag}" "${flags_with_completion[@]}"
                if [[ ${index} -ge 0 ]]; then
                    COMPREPLY=()
                    cur="${cur#*=}"
                    ${flags_completion[${index}]}
                    if [ -n "${ZSH_VERSION}" ]; then
                        # zfs completion needs --flag= prefix
                        eval "COMPREPLY=( \"\${COMPREPLY[@]/#/${flag}=}\" )"
                    fi
                fi
            fi
            return 0;
            ;;
    esac

    # check if we are handling a flag with special work handling
    local index
    __index_of_word "${prev}" "${flags_with_completion[@]}"
    if [[ ${index} -ge 0 ]]; then
        ${flags_completion[${index}]}
        return
    fi

    # we are parsing a flag and don't have a special handler, no completion
    if [[ ${cur} != "${words[cword]}" ]]; then
        return
    fi

    local completions
    completions=("${commands[@]}")
    if [[ ${#must_have_one_noun[@]} -ne 0 ]]; then
        completions=("${must_have_one_noun[@]}")
    fi
    if [[ ${#must_have_one_flag[@]} -ne 0 ]]; then
        completions+=("${must_have_one_flag[@]}")
    fi
    COMPREPLY=( $(compgen -W "${completions[*]}" -- "$cur") )

    if [[ ${#COMPREPLY[@]} -eq 0 && ${#noun_aliases[@]} -gt 0 && ${#must_have_one_noun[@]} -ne 0 ]]; then
        COMPREPLY=( $(compgen -W "${noun_aliases[*]}" -- "$cur") )
    fi

    if [[ ${#COMPREPLY[@]} -eq 0 ]]; then
        declare -F __custom_func >/dev/null && __custom_func
    fi

    __ltrim_colon_completions "$cur"
}

# The arguments should be in the form "ext1|ext2|extn"
__handle_filename_extension_flag()
{
    local ext="$1"
    _filedir "@(${ext})"
}

__handle_subdirs_in_dir_flag()
{
    local dir="$1"
    pushd "${dir}" >/dev/null 2>&1 && _filedir -d && popd >/dev/null 2>&1
}

__handle_flag()
{
    __debug "${FUNCNAME[0]}: c is $c words[c] is ${words[c]}"

    # if a command required a flag, and we found it, unset must_have_one_flag()
    local flagname=${words[c]}
    local flagvalue
    # if the word contained an =
    if [[ ${words[c]} == *"="* ]]; then
        flagvalue=${flagname#*=} # take in as flagvalue after the =
        flagname=${flagname%%=*} # strip everything after the =
        flagname="${flagname}=" # but put the = back
    fi
    __debug "${FUNCNAME[0]}: looking for ${flagname}"
    if __contains_word "${flagname}" "${must_have_one_flag[@]}"; then
        must_have_one_flag=()
    fi

    # if you set a flag which only applies to this command, don't show subcommands
    if __contains_word "${flagname}" "${local_nonpersistent_flags[@]}"; then
      commands=()
    fi

    # keep flag value with flagname as flaghash
    if [ -n "${flagvalue}" ] ; then
        flaghash[${flagname}]=${flagvalue}
    elif [ -n "${words[ $((c+1)) ]}" ] ; then
        flaghash[${flagname}]=${words[ $((c+1)) ]}
    else
        flaghash[${flagname}]="true" # pad "true" for bool flag
    fi

    # skip the argument to a two word flag
    if __contains_word "${words[c]}" "${two_word_flags[@]}"; then
        c=$((c+1))
        # if we are looking for a flags value, don't show commands
        if [[ $c -eq $cword ]]; then
            commands=()
        fi
    fi

    c=$((c+1))

}

__handle_noun()
{
    __debug "${FUNCNAME[0]}: c is $c words[c] is ${words[c]}"

    if __contains_word "${words[c]}" "${must_have_one_noun[@]}"; then
        must_have_one_noun=()
    elif __contains_word "${words[c]}" "${noun_aliases[@]}"; then
        must_have_one_noun=()
    fi

    nouns+=("${words[c]}")
    c=$((c+1))
}

__handle_command()
{
    __debug "${FUNCNAME[0]}: c is $c words[c] is ${words[c]}"

    local next_command
    if [[ -n ${last_command} ]]; then
        next_command="_${last_command}_${words[c]//:/__}"
    else
        if [[ $c -eq 0 ]]; then
            next_command="_$(basename "${words[c]//:/__}")"
        else
            next_command="_${words[c]//:/__}"
        fi
    fi
    c=$((c+1))
    __debug "${FUNCNAME[0]}: looking for ${next_command}"
    declare -F "$next_command" >/dev/null && "$next_command"
}

__handle_word()
{
    if [[ $c -ge $cword ]]; then
        __handle_reply
        return
    fi
    __debug "${FUNCNAME[0]}: c is $c words[c] is ${words[c]}"
    if [[ "${words[c]}" == -* ]]; then
        __handle_flag
    elif __contains_word "${words[c]}" "${commands[@]}"; then
        __handle_command
    elif [[ $c -eq 0 ]] && __contains_word "$(basename "${words[c]}")" "${commands[@]}"; then
        __handle_command
    else
        __handle_noun
    fi
    __handle_word
}

_heketi-cli_cluster_create()
{
    last_command="heketi-cli_cluster_create"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_cluster_delete()
{
    last_command="heketi-cli_cluster_delete"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_cluster_info()
{
    last_command="heketi-cli_cluster_info"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_cluster_list()
{
    last_command="heketi-cli_cluster_list"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_cluster()
{
    last_command="heketi-cli_cluster"
    commands=()
    commands+=("create")
    commands+=("delete")
    commands+=("info")
    commands+=("list")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device_add()
{
    last_command="heketi-cli_device_add"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--name=")
    local_nonpersistent_flags+=("--name=")
    flags+=("--node=")
    local_nonpersistent_flags+=("--node=")
    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device_delete()
{
    last_command="heketi-cli_device_delete"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device_disable()
{
    last_command="heketi-cli_device_disable"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device_enable()
{
    last_command="heketi-cli_device_enable"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device_info()
{
    last_command="heketi-cli_device_info"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_device()
{
    last_command="heketi-cli_device"
    commands=()
    commands+=("add")
    commands+=("delete")
    commands+=("disable")
    commands+=("enable")
    commands+=("info")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node_add()
{
    last_command="heketi-cli_node_add"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--cluster=")
    local_nonpersistent_flags+=("--cluster=")
    flags+=("--management-host-name=")
    local_nonpersistent_flags+=("--management-host-name=")
    flags+=("--storage-host-name=")
    local_nonpersistent_flags+=("--storage-host-name=")
    flags+=("--zone=")
    local_nonpersistent_flags+=("--zone=")
    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node_delete()
{
    last_command="heketi-cli_node_delete"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node_disable()
{
    last_command="heketi-cli_node_disable"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node_enable()
{
    last_command="heketi-cli_node_enable"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node_info()
{
    last_command="heketi-cli_node_info"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_node()
{
    last_command="heketi-cli_node"
    commands=()
    commands+=("add")
    commands+=("delete")
    commands+=("disable")
    commands+=("enable")
    commands+=("info")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_setup-openshift-heketi-storage()
{
    last_command="heketi-cli_setup-openshift-heketi-storage"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--listfile=")
    local_nonpersistent_flags+=("--listfile=")
    flags+=("--durability=")
    local_nonpersistent_flags+=("--durability=")
    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_topology_info()
{
    last_command="heketi-cli_topology_info"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_topology_load()
{
    last_command="heketi-cli_topology_load"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json=")
    two_word_flags+=("-j")
    local_nonpersistent_flags+=("--json=")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_topology()
{
    last_command="heketi-cli_topology"
    commands=()
    commands+=("info")
    commands+=("load")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume_create()
{
    last_command="heketi-cli_volume_create"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--clusters=")
    local_nonpersistent_flags+=("--clusters=")
    flags+=("--disperse-data=")
    local_nonpersistent_flags+=("--disperse-data=")
    flags+=("--durability=")
    local_nonpersistent_flags+=("--durability=")
    flags+=("--name=")
    local_nonpersistent_flags+=("--name=")
    flags+=("--persistent-volume")
    local_nonpersistent_flags+=("--persistent-volume")
    flags+=("--persistent-volume-endpoint=")
    local_nonpersistent_flags+=("--persistent-volume-endpoint=")
    flags+=("--persistent-volume-file=")
    local_nonpersistent_flags+=("--persistent-volume-file=")
    flags+=("--redundancy=")
    local_nonpersistent_flags+=("--redundancy=")
    flags+=("--replica=")
    local_nonpersistent_flags+=("--replica=")
    flags+=("--size=")
    local_nonpersistent_flags+=("--size=")
    flags+=("--snapshot-factor=")
    local_nonpersistent_flags+=("--snapshot-factor=")
    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume_delete()
{
    last_command="heketi-cli_volume_delete"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume_expand()
{
    last_command="heketi-cli_volume_expand"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--expand-size=")
    local_nonpersistent_flags+=("--expand-size=")
    flags+=("--volume=")
    local_nonpersistent_flags+=("--volume=")
    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume_info()
{
    last_command="heketi-cli_volume_info"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume_list()
{
    last_command="heketi-cli_volume_list"
    commands=()

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli_volume()
{
    last_command="heketi-cli_volume"
    commands=()
    commands+=("create")
    commands+=("delete")
    commands+=("expand")
    commands+=("info")
    commands+=("list")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

_heketi-cli()
{
    last_command="heketi-cli"
    commands=()
    commands+=("cluster")
    commands+=("device")
    commands+=("node")
    commands+=("setup-openshift-heketi-storage")
    commands+=("topology")
    commands+=("volume")

    flags=()
    two_word_flags=()
    local_nonpersistent_flags=()
    flags_with_completion=()
    flags_completion=()

    flags+=("--json")
    flags+=("--log-flush-frequency=")
    flags+=("--secret=")
    flags+=("--server=")
    two_word_flags+=("-s")
    flags+=("--user=")
    flags+=("--version")
    flags+=("-v")
    local_nonpersistent_flags+=("--version")

    must_have_one_flag=()
    must_have_one_noun=()
    noun_aliases=()
}

__start_heketi-cli()
{
    local cur prev words cword
    # TODO: What is the use of flaghash?
    # It seems it's filled but never read?!
    # shellcheck disable=SC2034
    declare -A flaghash 2>/dev/null || :
    if declare -F _init_completion >/dev/null 2>&1; then
        _init_completion -s || return
    else
        __my_init_completion -n "=" || return
    fi

    local c=0
    local flags=()
    local two_word_flags=()
    local local_nonpersistent_flags=()
    local flags_with_completion=()
    local flags_completion=()
    local commands=("heketi-cli")
    local must_have_one_flag=()
    local must_have_one_noun=()
    local last_command
    local nouns=()

    __handle_word
}

if [[ $(type -t compopt) = "builtin" ]]; then
    complete -o default -F __start_heketi-cli heketi-cli
else
    complete -o default -o nospace -F __start_heketi-cli heketi-cli
fi

# ex: ts=4 sw=4 et filetype=sh
